### PR TITLE
✨ Noise-Aware Simulator Refactor and Backends

### LINKED ISSUE #89: Expanding support of non-unitary functions for the deterministic quantum circuit simulator
Adding measurement, classical controlled operations, and reset operations to the density matrix simulator.

### LINKED ISSUE #93: Expanding support of non-unitary functions for the stochastic quantum circuit simulator
The stochastic noise aware simulator currently does not support reset operations. Code for that is already available here
https://github.com/cda-tum/ddsim/blob/f4459934be255d08b30df9a60bde041f0d1bdb8a/include/StochasticNoiseSimulator.hpp
https://github.com/cda-tum/ddsim/blob/f4459934be255d08b30df9a60bde041f0d1bdb8a/src/StochasticNoiseSimulator.cpp

The code must be implemented again and tested.

### PR DESCRIPTION
**Pull request by 33Gjl1Xe**
## Description

This PR adds support for non-unitary operations such as mid-circuit measurements and resets to the noise-aware simulation approaches in DDSIM.
Furthermore, it exposes both noise-aware simulators as Qiskit backends. For the first time, this makes them available to the Python users of DDSIM.

Fixes #89
Fixes #93

## Checklist:

- [x] The pull request only contains commits that are related to it.
- [x] I have added appropriate tests and documentation.
- [x] I have made sure that all CI jobs on GitHub pass.
- [x] The pull request introduces no new warnings and follows the project's style guidelines.

### CODE REVIEW

File: src/mqt/ddsim/stochasticnoisesimulator.py
Comment (burgholzer) [inline src/mqt/ddsim/stochasticnoisesimulator.py:22]:
```suggestion
            description="MQT DDSIM noise-aware stochastic simulator based on decision diagrams",
```

File: src/mqt/ddsim/deterministicnoisesimulator.py
Comment (burgholzer) [inline src/mqt/ddsim/deterministicnoisesimulator.py:21]:
```suggestion
            name="density_matrix_dd_simulator",
```
Just a little shorter name that is still concise.

File: src/mqt/ddsim/deterministicnoisesimulator.py
Comment (burgholzer) [inline src/mqt/ddsim/deterministicnoisesimulator.py:22]:
```suggestion
            description="MQT DDSIM noise-aware density matrix simulator based on decision diagrams",
```

File: test/python/noiseawareSimulator/test_noiseaware_determinstic_simulator.py
Comment (burgholzer) [inline test/python/noiseawareSimulator/test_noiseaware_determinstic_simulator.py:16]: Please try to remove unused/commented-out code.

**Review by burgholzer — CHANGES_REQUESTED** [review]
Many thanks also here! This is looking great already as it simplifies quite some code and finally exposes the noise-aware simulators to Python.
As for the Python side, I believe only minimal changes are needed to make this work. The biggest probably being adding some tests for the new Qiskit backends.

On the C++ side, I also added some comments that might at least require a little bit of work. There my main concern is that there is quite some duplication of code between the various files where I really believe that it shouldn't be too hard to unify the implementations.
Overall, the code is really looking good already and I really like the `analyseCircuit` method. I might even pull that to mqt-core at some point.

Let me know if you have any questions 👍🏼

**Comment by burgholzer** [comment]
I just merged #348 which contains a minimal set of changes required to make DDSIM compatible with the latest changes in mqt-core. Naturally, this causes some conflicts here, that need to be resolved. However, given that it is yet unclear on how fast we can get this in, I figured it is better to unblock the main branch from changes as quickly as possible and take the slight hit here.

File: src/DeterministicNoiseSimulator.cpp
Comment (33Gjl1Xe) [reply src/DeterministicNoiseSimulator.cpp]: Refactored that code to avoid code duplication.

File: src/DeterministicNoiseSimulator.cpp
Comment (33Gjl1Xe) [reply src/DeterministicNoiseSimulator.cpp]: Refactored that code to avoid code duplication.

File: src/DeterministicNoiseSimulator.cpp
Comment (33Gjl1Xe) [reply src/DeterministicNoiseSimulator.cpp]: Refactored that code to avoid code duplication.

File: include/DeterministicNoiseSimulator.hpp
Comment (33Gjl1Xe) [reply include/DeterministicNoiseSimulator.hpp]: I am overriding the function "measureAllNonCollapsing" now.

File: src/StochasticNoiseSimulator.cpp
Comment (33Gjl1Xe) [reply src/StochasticNoiseSimulator.cpp:72]: The variable is used in line 158. Yet, my IDE marks the variable as unused. To avoid the warning I marked the variable maybe_unused.

File: src/python/bindings.cpp
Comment (33Gjl1Xe) [reply src/python/bindings.cpp:59]: I changed the code as suggested.

File: src/mqt/ddsim/deterministicnoisesimulator.py
Comment (33Gjl1Xe) [reply src/mqt/ddsim/deterministicnoisesimulator.py:14]: I changed the name to DNS (=Deterministic_Noise-aware_Simulator)

File: src/mqt/ddsim/stochasticnoisesimulator.py
Comment (33Gjl1Xe) [reply src/mqt/ddsim/stochasticnoisesimulator.py:21]: I changed the name a suggested.

File: src/mqt/ddsim/deterministicnoisesimulator.py
Comment (33Gjl1Xe) [reply src/mqt/ddsim/deterministicnoisesimulator.py:21]: I changed the name as suggested.

File: src/mqt/ddsim/deterministicnoisesimulator.py
Comment (33Gjl1Xe) [reply src/mqt/ddsim/deterministicnoisesimulator.py:22]: I changed the name as suggested.

File: src/mqt/ddsim/stochasticnoisesimulator.py
Comment (33Gjl1Xe) [reply src/mqt/ddsim/stochasticnoisesimulator.py:1]: I added the backends as suggested.

**Review by burgholzer — APPROVED** [review]
@33Gjl1Xe Many thanks for your work on this!

I took the liberty to refactor the overall functionality some more and fix certain aspects. At this point, I think the PR is ready to merge as soon as all checks pass.

I believe there are a couple of nice follow-ups after that:
- In #347, the documentation for the simulators has been enhanced quite a bit. It would be nice to transform the existing documentation on the noise-aware simulators in a similar fashion (i.e., in the form of a Jupyter notebook)
- Qiskit has its own way of specifying noise models for simulators. It would be great if we would support those models directly instead of our backends having their own options.
- The repository could really use some more tests for these new simulators. Especially the support for dynamic circuits.

Neither of the above needs to happen in this PR and would only prolong the time until this gets merged. They would be much appreciated as a follow-up though. (@33Gjl1Xe 😉)

**Pull request merged**
